Remove XHTML for a More Consistent HTML5 Markup

DangitRick opened the pull request:
## Description
This pull request includes many small changes intended to remove XHTML markup in favor of HTML5 equivalents. 

## Motivation and context
HTML5 became the "next standard" for markup on the web back in 2007, with XHTML being obsolesced. By removing XHTML code from core, we are free to move forward with full HTML5 markup and practices, with more consistent code and less PHP to choose between markup alternatives.

It also gets us closer to pages with clean markup that validates on the W3C's validator.

## What's Left to Do?
I don't know JavaScript, so I did not modify any JS files, though many showed up when searching for "/>" for example.

Posts still output closing slashes, such as `<hr />`, when self-closing elements are used.

I also avoided touching files that were obviously part of third party packages, such as SimplePie or PHPmailer.

## How has this been tested?
On both my local and live installations, everything appears to be working smoothly. Using a test site with content from FakerPress, the W3C validator's results went from a couple dozen errors and warnings down to just a few.

## Types of changes

* Remove closing slashes from self-closing elements
* Remove double-barreled boolean attributes in favor of their simpler forms (e.g., prefer `selected` over `selected="selected"`
* Remove `current_theme_supports('html5')` and all logic related to it in favor of just using HTML5
* Remove `text/css` and `text/javascript` `type` attributes from `script` and `link` tags. 
* Remove assorted `role` attributes on HTML5 elements such as `nav` or `header` where the role is already implied by the markup
* Remove `novalidate` attributes from forms to enable client-side browser-based validation, providing users with instant form feedback while saving on system resources

DangitRick committed:
wp-signup.php XHTML cleanup

Removed closing slashes from self-closing tags.

Removed `type="text/css"` from `style` tag.

DangitRick committed:
TwentyFifteen customize-preview.js XHTML cleanup

Removed `/` from self-closing tag.

Removed `type="text/css"` from `style` tag.

DangitRick committed:
TwentySixteen searchform.php XHTML removal

Removed closing slash from self-closing tag.

DangitRick committed:
TwentySixteen customize-preview.js - Remove XHTML

Removed `type="text/css"` from `style` tag.

Removed closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from twentyseventeen searchform.php

Remove closing slash from self-closing tag

DangitRick committed:
Remove XHTML from deprecated.php

Remove trailing slashes from self-closing tags

DangitRick committed:
Remove XHTML from wp-login.php

Remove `type="text/javascript"` from `script` tags.

Remove closing slashes from self-closing tags.

DangitRick committed:
Remove XHTML from author-template.php

Remove closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from class-wp-customize-date-time-control.php

Remove closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from class-wp-customize-color-control.php

Remove closing slashes from self-closing tags.

DangitRick committed:
Remove XHTML from class-wp-customize-nav-menu-locations-control

Remove trailing slash from self-closing tag.

DangitRick committed:
Remove XHTML from class-wp-customize-site-icon-control.php

Remove closing slash from self-closing tags.

DangitRick committed:
Remove XHTML from class-wp-customize-themes-section.php

Remove closing slashes from self-closing tags.

DangitRick committed:
Remove XHTML from class-wp-customize-nav-menu-auto-add-control.php

Remove closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from class-wp-customize-theme-control.php

Remove closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from general-template.php

Remove closing slashes from self-closing elements.

DangitRick committed:
Update get_the_generator()

Remove `xhtml` as an option; default to `html`.

DangitRick committed:
Remove XHTML search form option

Default `get_search_form()` to outputting HTML5 markup, removing the XHTML option.

DangitRick committed:
Remove 'xhtml' as an option in header language tags

Removed the logic for choosing between 'html' and 'xhtml' in the language tags.

DangitRick committed:
Simplify boolean attributes

Simplify boolean attributes throughout core. (Based on attributes list here <https://meiert.com/en/blog/boolean-attributes-of-html/>)

Did not touch JS files or files in included third-party packages, like SimplePie.

DangitRick committed:
Enable Client-Side Form Validation

Removes `novalidate` attributes from all forms to allow client-side validation of such as email addresses.

DangitRick committed:
Default Atom feeds to `html` type

If ClassicPress doesn't do XHTML, then the Atom feeds will never output XHTML content. Enforce 'html'.

DangitRick committed:
Remove XHTML from class-wp-widget-rss.php

Remove choice between xhtml and html to default to html.

Remove closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from class-wp-nav-menu-widget.php

Remove choice between xhtml and html to default to html.

Remove closing slash from self-closing tag.

DangitRick committed:
Remove XHTML from class-wp-widget-recent-comments.php

Remove choice between xhtml and html to default to html.

Remove closing slashes on self-closing tags.

DangitRick committed:
Remove XHTML from comments system.

Removed `comment()` function in favor of `html5_comment()`.

Remove closing slashes from self-closing tags.

DangitRick committed:
Remove XHTML from class-wp-widget-tag-cloud.php

Remove xhtml/html choice.

Remove closing slashes on self-closinig elements.

DangitRick committed:
Remove XHTML from class-wp-widget-categories.php

Remove xhtml/html choice to default to HTML5.

Remove closing slashes on self-closing tags.

DangitRick committed:
Remove XHTML from class-wp-widget-archives.php

Remove XHTML/HTML choice.

Remove JavaScript `type` attribute.

Remove closing slashes from self-closing tags.

DangitRick committed:
Default twentyfifteen to HTML5

No need to explicitly declare HTML5 support.

DangitRick committed:
Default twentysixteen to HTML5

No need to explicitely declare support for HTML5.

DangitRick committed:
Default twentyseventeen to HTML5.

No need to explicitly declare support for HTML5.

DangitRick committed:
Remove XHTML from script-loader.php

Remove CDATA markup.

Remove JavaScript an Style `type` attributes.

Remove closing slashes from self-closing tags.

DangitRick committed:
Lots of miscellaneous XHTML removals.

DangitRick committed:
More XHTML removal.

DangitRick committed:
Origin pull

DangitRick committed:
More XHTML cleanup. (It never ends!)

DangitRick committed:
Further XHTML cleanup.

DangitRick committed:
Further XHTML cleanup.

DangitRick committed:
More XHTML cleanup

Lots of slashes removed.

role="navigation" is not require on NAV elements.

DangitRick committed:
role="main" not require on <main> elements

DangitRick committed:
role="complimentary" not required on <aside> tags

DangitRick committed:
role="contentinfo" not required on footer elements

DangitRick committed:
More XHTML excision

DangitRick committed:
Revert "Origin pull"

This reverts commit bf85689eebfac682ab44ad740d42bb87c106689d.

DangitRick commented:
Not mentioned above: 
* Removed CDATA from JavaScript blocks (not needed in HTML5)

And for under the To Do:
It may be wise to enable HTML5 theme support by default for any plugins that may do an XHTML/HTML5 check. I don't think themes should be required to explicitly declare the support if it's meant to be default at this point. I wasn't sure of the best way to do this in core, though.

DangitRick committed:
Resolve a conflict.

DangitRick committed:
Resolve a conflict.

DangitRick committed:
Resolve a conflict

Also, change XHTML to HTML

DangitRick committed:
Resolve conflict.

DangitRick committed:
Conflict resolve?

xxsimoxx commented:
Just got a quick check on the PR.
What about functions like [__checked_selected_helper](https://github.com/ClassicPress/ClassicPress-v2/blob/d869dcc93374e60053a028c03089abe18b0b8cab/src/wp-includes/general-template.php#L5122)?

DangitRick committed:
Fixes a helper function that output doubled HTML5 attributes.

DangitRick committed:
Added @since comment.

DangitRick committed:
Merge branch 'develop' into xhtml-excision

DangitRick commented:
Fixed marked conflicts and fixed the `__checked_selected_helper()` function.

DangitRick committed:
Fix PHP coding standards

DangitRick committed:
Merge branch 'xhtml-excision' of https://github.com/DangitRick/ClassicPress-v2 into xhtml-excision

DangitRick committed:
PHP standards

DangitRick commented:
Been having a hard time getting a local environment set up to test this locally, per the request made on Zulip. When running `composer run phpcs` I was getting a ton of errors until I silenced the `trim()` function on line 194 of `vendor/wp-coding-standards/wpcs/WordPress/Sniffs/WP/i18nSniff.php` by adding an `@` in front of it. 

After that, the tests pass 20/20. 

Running `phpunit` tests isn't getting anywhere, though; every time, no matter what I do, I get a "cannot connect to database" error, despite the connection working fine when browsing to the site via localhost. 

Does anyone have any idea on what the best way to permanently mark `html5` as being enabled in theme support, for maximum plugins capacity? Or how about modifying TinyMCU output so that no closing-slashes are output into post content?

mattyrob commented:
I think the `trim()` issues may be due to the PHP version you are using locally.

Unit tests may need configuration of a local `wp-tests-config.php` file from the included `wp-tests-config-sample.php` file.

I was wondering if we need to re-write `current_theme_supports()` in `src/wp-includes/theme.php` a little such that is always returned `true` for any call to `current_theme_supports( 'html5' );`.

TinyMCE may need some more investigation and perhaps a plugin or a filter on the content to remove trailing slashes within the editor.

xxsimoxx commented:
About TinyMCE, this code does the trick.

```php
add_filter(
	'tiny_mce_before_init',
	function ( $config ) {
		$config['element_format'] = 'html';
		return $config;
	}
);
```

mattyrob commented:
We need to think about where that MCE config code should go.
It seems there is already a filter applied to that in `src/wp-includes/default-filters.php` with the function contained in `src/wp-includes/functions.php`.

I wonder though is we can add the hook in `src/wp-admin/includes/admin-filters.php` if it is only needed in admin and maybe place the function itself in `src/wp-includes/classicpress/class-cp-customization.php`

mattyrob commented on src/wp-includes/media-template.php:
```diff
@@ -584,16 +584,16 @@ function wp_print_media_templates() {
+							<img src="{{ data.image.src }}" class="thumbnail" draggable="false" alt="">
 						<# } else if ( data.sizes && data.sizes.medium ) { #>
-							<img src="{{ data.sizes.medium.url }}" class="thumbnail" draggable="false" alt="" />
-						<# } else { #>
```
This is incorrect, the ` #>` needs restoring here to prevent JS errors in the Media Library.

xxsimoxx commented:
> It seems there is already a filter applied to that in `src/wp-includes/default-filters.php` with the function contained in `src/wp-includes/functions.php`.

In this way if a site enables tinyMCE for comments we get the right tags, not just in the post editor, so for me it is the best choice. 

Edit: [here](https://github.com/DangitRick/ClassicPress-v2/compare/xhtml-excision...xxsimoxx:ClassicPress-v2:fix/tiny) the code.

DangitRick committed:
incorporate suggested changes

Fixed broken syntax in media library

Add code to get TinyMCE to output HTML

DangitRick commented:
Fixed the above issue from @mattyrob and added the TinyMCE fix (dang, it was simpler than I thought!) from @xxsimoxx.

Local phpcs test returns no errors. Local phpunit test returns with 11073 errors and 122 warnings… Everything works on the test install, though. Version of PHP used: 8.2.0.

xxsimoxx commented:
@DangitRick can't see the new commits for TinyMCE and the template issue.
When I run `phpunit` I get about 240 errors, 10k seems to be a local issue.

viktorix commented on src/wp-includes/functions.php:
```diff
@@ -4652,6 +4652,22 @@ function _mce_set_direction( $mce_init ) {
 }
 
+/**
+ * Sets elemenmt format to html.
```
"element"

mattyrob committed:
Merge branch 'develop' into pr/200

mattyrob committed:
Merge branch 'xhtml-excision' of https://github.com/DangitRick/ClassicPress-v2 into pr/200

mattyrob committed:
Fix typo

mattyrob committed:
Restore some necessary core code

mattyrob committed:
First pass fixing tests

mattyrob committed:
Fix test coding standards

xxsimoxx reviewed (approved):
Looks good to me.

DangitRick committed:
Merge branch 'develop' into xhtml-excision

DangitRick committed:
Fix spelling mistake (thanks @viktorix

DangitRick committed:
Merge branch 'xhtml-excision' of https://github.com/DangitRick/ClassicPress-v2 into xhtml-excision

mattyrob committed:
Fixes for coding standards

mattyrob commented:
I think this is very close to merge now.

The last few things as far as I know is:
`current_theme_supports( 'html5' )` needs to return true always
`add_theme_support( 'html5' )` needs to return a `_doing_it_wrong()` with an updated message.
Tests for the above need updating.

I am working on this now.

mattyrob committed:
Update html5 theme support functions and tests

mattyrob committed:
Coding standards correction

mattyrob committed:
Remove unnecessary slashes in get_the_generator()

mattyrob committed:
Merge branch 'develop' into xhtml-excision
